feat: add includeUntested option to expose coverage of all instrumented files

marco-ramirez opened the pull request:
This allows reporting on all instrumented files even if they are not `require`d.

bcoe commented:
@marco-ramirez I'd like to better understand the problem that this solves; is the goal to show all files in the project, including those that have not been required.

Currently this can be achieved by using `nyc`'s `--all` setting, in conjunction with ` babel-plugin-istanbul`... I do like the idea that we could better support this in babel-plugin-istanbul itself though. 

What test suites are you using this approach with?

marco-ramirez commented:
@bcoe that is exactly the problem I am trying to solve.  I have a node project that doesn't use `nyc`, just gulp and mocha.  I'd like to do the following:
```
import gulp from 'gulp';
import babel from 'gulp-babel';
import mocha from 'gulp-mocha';
import istanbul from 'gulp-istanbul';

const babelConfig = {
    presets: ['es2015'],
    plugins: [
        ['istanbul', {includeUntested: true}]
    ]
};

gulp.task('instrument', () =>
    gulp.src('src/**.js')
        .pipe(babel(babelConfig))
        .pipe(istanbul.hookRequire()));

gulp.task('test', ['instrument'], () =>
    gulp.src('test/**/*.js')
        .pipe(mocha())
        .pipe(istanbul.writeReports({coverageVariable: '__coverage__'})));
```

bcoe commented:
@marco-ramirez it's a little bit weird to use `gulp-istanbul` and `babel-plugin-istanbul` in conjunction; The two libraries use two different versions of the same istanbul instrumentation logic.

I think rather than adding additional logic to babel-plugin-istanbul`, you should be able to get away with the approach outlined here:

https://istanbul.js.org/docs/tutorials/es2015/

and setting `all = true`. Is the project you're instrumenting open-source, or do you have a similar project that we could test out this approach on?

marco-ramirez commented:
I can achieve the same result as my previous example without using gulp like so:
```js
import glob from 'glob';
import {transform} from 'babel-core';
import fs from 'fs';
import {hookRequire} from 'istanbul-lib-hook';
import Mocha from 'mocha';
import path from 'path';
import {Collector, Reporter} from 'istanbul';

const babelOptions = {
    presets: ['es2015'],
    plugins: [
        ['istanbul', {includeUntested: true}]
    ]
};

const report = () => {
    const reporter = new Reporter();
    const collector = new Collector();
    reporter.add('text');
    collector.add(global.__coverage__);
    reporter.write(collector, false, () => {});
};

const runTests = () => {
    glob('test/**/*.js', (error, files) => {
        const mocha = new Mocha();
        files.forEach(file => mocha.addFile(file));
        mocha.run().on('end', report);
    });
};

glob('src/**/*.js', (error, files) => {
    const instrument = filename =>
        transform(fs.readFileSync(filename, {encoding: 'utf8'}), {...babelOptions, filename}).code;
    const cache = files.reduce((map, file) => {
        const filePath = path.resolve(file);
        return {...map, [filePath]: instrument(filePath)};
    }, {});
    const matcher = file => cache.hasOwnProperty(file);
    const transformer = (code, file) => cache[file];
    hookRequire(matcher, transformer);
    runTests();
});
```
That I can instrument source files for code coverage and generate coverage reports so easily is a testament to the utility of istanbul and babel.  It is my belief that adding this feature to the plugin simplifies its integration into other tools and workflows. If this plugin is only intended to support `nyc`, then you're absolutely right that this feature is redundant.

bcoe commented:
@marco-ramirez sorry for the slow reply (I pretty much only have time for OSS on weekends these days); 

Sorry to ask for the annoying extra work, but would it be possible to provide a testing repo, [something like this](https://github.com/wbyoung/nyc-issue-482-example), showing your new API in action. I think this will help me better visualize the value of this addition (and it also provides some great documentation for us to point folks at) 👍 

thanks for the hard work so far.

marco-ramirez commented:
@bcoe don't worry about your reply speed. I'm in the same boat. Here's a [sample repo](https://github.com/marco-ramirez/babel-plugin-istanbul-include-untested-example).

novemberborn commented:
I find it strange how the Babel plugin modifies a global in the process that is executing the plugin itself. This must assume the plugin is used with a call to `babel.transform()` rather than through `babel-cli`. Perhaps instead the plugin could support a callback option which is called with the `this.__dv__.exit(path)` result?

```js
const coverage = {}

const babelOptions = {
    presets: ['es2015'],
    plugins: [
        ['istanbul', {onCover (file, result) { coverage[file] = result.fileCoverage }]
    ]
};

const report = () => {
    const reporter = new Reporter();
    const collector = new Collector();
    reporter.add('text');
    collector.add(coverage);
    reporter.write(collector, false, () => {});
};
```

And:

```js
let result = this.__dv__.exit(path)
if (this.opts.onCover) {
  this.opts.onCover(getRealpath(this.file.opts.filename), result)
}
```

IIRC Babel does not validate plugin options so this should work.

bcoe commented:
@marco-ramirez I like @novemberborn's recommendation. Feels like a cleaner way to output the coverage information to the context of your gulp configuration. If I understand @novemberborn's suggestion appropriately:

we would always be calling `onCover` if the closure is provided? (whether files have been instrumented or not? I believe this would remove the need for the `includeUntested` flag?

marco-ramirez commented:
@bcoe I implemented the `onCover` callback as recommended by @novemberborn. I also updated the [usage example](https://github.com/marco-ramirez/babel-plugin-istanbul-include-untested-example).  The only thing I had do differently than what was suggested was to add both the global `__coverage__` object and the coverage object built from the onCover callback to the collector.  Come to think of it, I could have just populated the global `__coverage__` object right from the callback.

bcoe commented on src/index.js:
```diff
@@ -57,7 +57,10 @@ function makeVisitor ({types: t}) {
             return
           }
-          this.__dv__.exit(path)
+          let result = this.__dv__.exit(path)
```
I think this could be a `const` since you don't re-assign the value.

marco-ramirez commented:
@bcoe I changed the `let` to `const`, and yes, it works like I need it to.

bcoe commented:
@marco-ramirez sorry that this took me so long to land! got buried in other OSS work;

Could you try running `npm i babel-plugin-istanbul@next` and let me know if this does the trick for you; if  so I'll promote the release.
